src/cargo/util/config/target.rs

```diff
@@ -64,18 +64,62 @@ pub(super) fn load_target_cfgs(config: &Config) -> CargoResult<Vec<(String, Targ
     Ok(result)
 }
 
+/// Returns true if the `[target]` table should be applied to host targets.
+pub(super) fn get_target_applies_to_host(config: &Config) -> CargoResult<bool> {
+    if config.cli_unstable().target_applies_to_host {
```

Comment:
Can an entry in the unstable book get added for this?

(personally I think it's fine if this is gated by `-Zhost-config` too)

Reply:
> Can an entry in the unstable book get added for this?

Added.

> (personally I think it's fine if this is gated by `-Zhost-config` too)

Well we require `-Ztarget-applies-to-host` to be enabled if `-Zhost-config` is enabled to ensure we appropriately test the stabilization ordering at the moment since we plan to stabilize the `target-applies-to-host` option before the `host-config` feature.

Reply:
Can you detail more why these should be stabilized separately? I would imagine that there's no need for them to be separate and they should be stabilized at the same time. It seems like it would be odd if we provide a way to configure host/target separately but then we don't provide a way to configure the host.

Reply:
Reason is that we don't want to stabilize `host-config` without the new defaults since that complicates testing logic, so we first stabilize `target-applies-to-host` so that everyone can choose what behavior they want(retain existing or use new) ahead of the default change. For common use cases the host tables aren't actually needed so `target-applies-to-host` should usually be sufficient anyways to make cross compilation usable with cargo in most cases.

Reply:
@alexcrichton This came out of discussions about the path to future stabilization, and the path to changing the defaults, and how those will interlock.

Reply:
(as a side note can you please stop marking all conversations resolved after you reply? I have to go hunting for the box that has the comments and it's kind of annoying, more generally as the one with the concern I'm probably the one to decide whether it's resolved or not...)

Personally I'd like to continue to dig in here. I continue to not understand why there needs to be two feature flags. Why would we want to stabilize `target-applies-to-host` before `host-config`? I do not thing that the `[host]` table has any more thorny of questions than `target-applies-to-host` does, so it does not seem to me like there's any need to have different stabilization schedules for them.

AFAIK the only thing that needs to be "delayed" or on a different schedule is switching the default value of `target-applies-to-host`, which would happen on nightly once `target-applies-to-host` reaches stable.

Reply:
> as a side note can you please stop marking all conversations resolved after you reply? I have to go hunting for the box that has the comments and it's kind of annoying, more generally as the one with the concern I'm probably the one to decide whether it's resolved or not...

Ok, I'll leave them open, I assumed you'd just reopen any that needed additional info.

> Why would we want to stabilize `target-applies-to-host` before `host-config`?

Because the `host-config` implementation complexity goes up significantly without the new defaults, it's a good bit harder to test as well.

> I do not thing that the `[host]` table has any more thorny of questions than `target-applies-to-host` does, so it does not seem to me like there's any need to have different stabilization schedules for them.

The `[host]` table behavior is affected by the defaults while the `target-applies-to-host` flag is not as the `target-applies-to-host` flag unconditionally overrides any defaults essentially.

> AFAIK the only thing that needs to be "delayed" or on a different schedule is switching the default value of `target-applies-to-host`, which would happen on nightly once `target-applies-to-host` reaches stable.

We don't want to allow a `[host]` table with the old defaults essentially, hence the delay.

Reply:
I continue basically remain unconvinced and not really understanding your rationale for splitting these two. 

For example I'm not sure what you mean as implementation complexity? As you and others have pointed out a few times, this is a small PR, and I would also agree that this is not a complex PR and doesn't really add a ton of implementation complexity. There is still a remaining problem of ensuring `cargo build` is handled correctly because it behaves differently from `cargo build --target $host`, and that *may* be complex. I think that issue is required to stabilize either option here, though, so I don't see how implementation complexity factors in.

I understand that `[host]` and `target-applies-to-host` interact with each other, but I don't understand why that explains the two feature gates. I'm proposing that there's just one feature gate that the behavior is gated on (both reading `target-applies-to-host` and the `[host]` table).

Basically I'm trying to understand the rationale for these being two separate feature gates. I understand what the two feature gates do, and I understand how they interact with each other. What I do not understand is why there are two feature gates instead of one.

Reply:
> There is still a remaining problem of ensuring `cargo build` is handled correctly because it behaves differently from `cargo build --target $host`, and that may be complex.

These shouldn't result in different behavior in my current implementation.

Reply:
Sorry it sounds like I'm advocating for something where Cargo detects the presence of `[host]` and acts on that as a feature gate of some form. Instead what I mean is that there's a singular `-Z` flag which gates these two new features in Cargo, the `target-applies-to-host` configuration option and `[host]` configuration table. Both of those are completely ignored if the `-Z` flag isn't passed. If the `-Z` flag is passed then they're both respected as if both were stable as-is in this PR. The default of `target-applies-to-host`, regardless of the `-Z` flag, will be `true`. Once this is all stable we'll flip that to `false`.

I understand there are literally two features here that are separated in the code. I am asking why, at an interface level, we are bothering to give users two different knobs. I don't think that the knobs are going to be stable at different times and otherwise only the "all on" and "all off" knob configurations are what we'd like to get feedback on feature-wise.

Reply:
> Instead what I mean is that there's a singular -Z flag which gates these two new features in Cargo, the target-applies-to-host configuration option and [host] configuration table. Both of those are completely ignored if the -Z flag isn't passed. If the -Z flag is passed then they're both respected as if both were stable as-is in this PR. The default of target-applies-to-host, regardless of the -Z flag, will be true. Once this is all stable we'll flip that to false.

The issue is that the implementation for one of these features(the `[host]` table feature) depends on the new defaults effectively. The -Z flag for `target-applies-to-host` does not change the defaults, however the -Z flag for `host-config` does as the host table implementation needs the new defaults.

> I understand there are literally two features here that are separated in the code. I am asking why, at an interface level, we are bothering to give users two different knobs. I don't think that the knobs are going to be stable at different times and otherwise only the "all on" and "all off" knob configurations are what we'd like to get feedback on feature-wise.

Well they are deliberately separated due to the dependency of the `host` table feature on the new defaults essentially, so they would be stabilized at different times as outlined [here](https://github.com/rust-lang/cargo/pull/9322#issuecomment-827293376). We have essentially one flag for the first step(`target-applies-to-host` setting) and one for the other(`host` table) with enforced ordering.

Reply:
I don't think `[host]` depends on new defaults, it just depends on `target-applies-to-host = false`. That can happen either with new defaults or the variable itself. Personally I don't think this is a reason to have two features.

To reiterate my thoughts from before, I disagree that these two features should be stabilized separately. They're pretty intertwined and all the hard stuff with one is still present for the other.

Reply:
> I don't think `[host]` depends on new defaults, it just depends on `target-applies-to-host = false`. That can happen either with new defaults or the variable itself.

Sure.

> Personally I don't think this is a reason to have two features.

That by itself isn't the reason for having the features separated, it's a testing issue mostly for both automated tests and those testing with the `-Z` flag, allowing the use of a `host-config` feature with the old defaults increases the combinations we need to test for and results in confusing/unclear behavior such as the host table being ignored if the old defaults are used(in the case that `target-applies-to-host` is not set before the default change). In addition we want to test the new default behavior as well so tying the new defaults to the `host-table` feature flag seems logical there as then either the `[host]` table or the new defaults can be tested with the `-Z` `host-config` flag together.

> To reiterate my thoughts from before, I disagree that these two features should be stabilized separately. They're pretty intertwined and all the hard stuff with one is still present for the other.

To some degree they are intertwined but it's mostly a one way dependency, `[host]` depends on `target-applies-to-host = false` but the opposite isn't really true, if we're going to change the defaults in the future it ends up being a lot easier/clearer to just tie that to the `[host]` table feature to simplify testing of the new defaults and avoid any unclear behavior for some combinations. Keep in mind also that the `[host]` table feature should only be needed for extremely exotic use cases, the `target-applies-to-host` feature should be sufficient to fix the broken builds in nearly all cases I can think of so I think it's fine delaying the `[host]` table feature stabilization to simplify the migration testing.

Reply:
As I've said many times before, I do not think it is useful to split up these features. I do not think that we can stabilize them on separate schedules, also as I've mentioned before.

This means, to me, the only two states of Cargo that we care about are both features on and both features off. Having one feature on and one off is not a useful state in Cargo since we'll never actually ship that to users. This, to me, means it's not worthwhile to test. It's testing functionality that no one should use, so I don't think we should even expose it and that would simplify things.

I'm honestly not really entirely certain why you're insistent on keeping these separate? As a maintainer of Cargo I'm offering my thoughts to help guide this implementation, but you seem very resolute in your ways that you want precisely the PR as-is right now. I don't really know how to reconcile that honestly. I've tried to respond to all the points you're bringing up but the discussion seems to be very circular where it's just coming around to the same points again and again.

Reply:
>  I do not think that we can stabilize them on separate schedules, also as I've mentioned before.

I guess I'm missing what the issue with this is I guess.

> This means, to me, the only two states of Cargo that we care about are both features on and both features off.

There's still the default change as well, which I also wanted to test, sure, we could combine the others and separate that if you want or just not test the new defaults I supposed but that seemed to me to result in a more confusing migration process as only allowing the host config table feature with the new defaults seem to simplify testing of that feature+default combination.

> Having one feature on and one off is not a useful state in Cargo since we'll never actually ship that to users. This, to me, means it's not worthwhile to test. It's testing functionality that no one should use, so I don't think we should even expose it and that would simplify things.

Well after the new defaults are applied one shouldn't need to set `target-applies-to-host = false` ever to use `host` tables, so I was just trying to avoid needing to set `target-applies-to-host = false` in cases where one also sets a host table as we probably won't have as good test coverage of all the combinations, we'll still probably be fine but I am less confident in regards to what the implementation correctness would be with them combined and having the default be introduced after the `host` table feature.

> I'm honestly not really entirely certain why you're insistent on keeping these separate? As a maintainer of Cargo I'm offering my thoughts to help guide this implementation, but you seem very resolute in your ways that you want precisely the PR as-is right now. I don't really know how to reconcile that honestly. I've tried to respond to all the points you're bringing up but the discussion seems to be very circular where it's just coming around to the same points again and again.

I'm just trying to clarify the issues I'm running into as it's unclear what the best approach to dealing with them in a satisfactory way is.

Well we inherently have two sets of changes either way, the `target-applies-to-host` feature needs to be introduced as part of the first, sure we could introduce a `host` table at the same time as that but we still have the default change down the line anyways so that's why I tied the new defaults to the `host` table to just have those both as part of the second set of changes.

It's unclear how you want me to implement the change though, is either of these approaches what you have in mind:

- Tie `target-applies-to-host` to `host` table feature and don't bother testing new defaults at all.
- Tie `target-applies-to-host` to `host` table feature and add a different `-Z` flag which enables the new defaults for testing.

I think the existing approach is less confusing but either of those would still work fine if that's what you think is best.